core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -230,30 +243,36 @@ object Scheduler {
       rest.prepend(Source.single(snapshot))
   }
 
-  // TODO (DCOS-47476) use actual prefixAndTail and expect first event to be a Snapshot; change the prefixAndTail param from 0 value to 1, let fail, etc.
-  private val stateOutputBreakoutFlow: Flow[StateEvent, StateOutput, NotUsed] = Flow[StateEvent].prefixAndTail(0).map {
-    case (_, stateEvents) =>
+  private val stateOutputBreakoutFlow: Flow[StateEvent, StateOutput, NotUsed] = Flow[StateEvent].prefixAndTail(1).map {
+    case (Seq(snapshot), stateEvents) =>
+      val stateSnapshot = snapshot match {
+        case x: StateSnapshot => x
+        case _ => throw new IllegalStateException("First event is allowed to be only a state snapshot")
+      }
       val stateUpdates = stateEvents.map {
         case _: StateSnapshot =>
           throw new IllegalStateException("Only the first event is allowed to be a state snapshot")
         case event => event
       }
-      (StateSnapshot.empty, stateUpdates)
+      (stateSnapshot, stateUpdates)
   }
 
-  def unconnectedGraph(
-      mesosCallFactory: MesosCalls): BidiFlow[SpecInput, StateOutput, MesosEvent, MesosCall, NotUsed] = {
+  private[core] def unconnectedGraph(
+      mesosCallFactory: MesosCalls,
+      podRecordRepository: PodRecordRepository): BidiFlow[SpecInput, StateOutput, MesosEvent, MesosCall, NotUsed] = {
+    val schedulerLogicGraph = new SchedulerLogicGraph(mesosCallFactory, loadPodRecords(podRecordRepository))
```

Comment:
This is getting closer to it but I don't understand why I cannot be a async. The two main factory methods `asFlow` and `asSinkAndSource` return futures. So you could simply call

```scala
  def asFlow(specsSnapshot: SpecsSnapshot, client: MesosClient)(
      implicit materializer: Materializer): Future[(StateSnapshot, Flow[SpecUpdated, StateEvent, NotUsed])] = async {

    implicit val ec = scala.concurrent.ExecutionContext.Implicits.global //only for ultra-fast non-blocking onComplete

    val initialRecords = await(loadRecords(repository))
    val (snap, source, sink) = asSourceAndSink(specsSnapshot, client, initialRecords)
    (await(snap), Flow.fromSinkAndSourceCoupled(sink, source))
}
```

Reply:
Discusses this offline, In an ideal world we should be using async-await construct to do this in a non-blocking fashion. But factory methods `fromFlow` and `fromClient` requires this to be done in a blocking fashion. I guess we can find a more cleaner solution once we finalize on a limited number of factory methods.

Based on the isolated fact that _we can't make any progress until we load the records_ I guess this is okay for now.

Reply:
`fromClient` and `fromFlow` could just be non-blocking. The signature of these methods is not set in stone.